streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java

```diff
@@ -0,0 +1,237 @@
+
+            final String threadIdPrefix = String.format("stream-thread [%s] ", Thread.currentThread().getName());
+            final String logPrefix = threadIdPrefix + String.format("%s [%s] ", "task", taskId);
+            final LogContext logContext = new LogContext(logPrefix);
+
+            final ProcessorTopology topology = builder.buildSubtopology(taskId.topicGroupId);
+
+            final ProcessorStateManager stateManager = new ProcessorStateManager(
+                taskId,
+                partitions,
+                Task.TaskType.ACTIVE,
+                stateDirectory,
+                topology.storeToChangelogTopic(),
+                storeChangelogReader,
+                logContext
+            );
+
+            if (threadProducer == null) {
+                // create one producer per task for EOS
+                // TODO: after KIP-447 this would be removed
+                final String taskProducerClientId = getTaskProducerClientId(threadId, taskId);
+                final Map<String, Object> producerConfigs = config.getProducerConfigs(taskProducerClientId);
+                producerConfigs.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, applicationId + "-" + taskId);
+                log.info("Creating producer client for task {}", taskId);
+                taskProducers.put(taskId, clientSupplier.getProducer(producerConfigs));
+            }
+
+            final RecordCollector recordCollector = new RecordCollectorImpl(
+                logContext,
+                taskId,
+                consumer,
+                threadProducer != null ?
+                    new StreamsProducer(threadProducer, false, logContext, applicationId) :
+                    new StreamsProducer(taskProducers.get(taskId), true, logContext, applicationId),
+                config.defaultProductionExceptionHandler(),
+                EXACTLY_ONCE.equals(config.getString(StreamsConfig.PROCESSING_GUARANTEE_CONFIG)),
+                streamsMetrics
+            );
+
+            final Task task = new StreamTask(
+                taskId,
+                partitions,
+                topology,
+                consumer,
+                config,
+                streamsMetrics,
+                stateDirectory,
+                cache,
+                time,
+                stateManager,
+                recordCollector
+            );
+
+            log.trace("Created task {} with assigned partitions {}", taskId, partitions);
+            createdTasks.add(task);
+            createTaskSensor.record();
+        }
+        return createdTasks;
+    }
+
+    public void releaseProducer() {
+        if (threadProducer != null) {
+            try {
+                threadProducer.close();
+            } catch (final RuntimeException e) {
+                log.error("Failed to close producer due to the following error:", e);
+            }
+        }
+        if (!taskProducers.isEmpty()) {
```

Comment:
Should we throw illegal state first, since we are already in an error state?

Reply:
If we're going to rename the method to specify that it should do exactly "close thread producer", then this check is no longer appropriate.